many: implement `snap refresh --list` and  `snap refresh`

mvo5 opened the pull request:
This is an early pull request mostly to get feedback on the actual rest api interface this should use.

chipaca commented on cmd/snap/cmd_list.go:
```diff
@@ -52,16 +53,31 @@ func (s snapsByName) Less(i, j int) bool { return s[i].Name < s[j].Name }
 	return listSnaps(x.Positional.Snaps)
 }
 
+func listUpdates() error {
```
can we s/updates/refreshes/ or something? this'll get confusing soon otherwise

chipaca commented on cmd/snap/cmd_snap_op.go:
```diff
@@ -183,24 +183,51 @@ type cmdRefresh struct {
-	name := x.Positional.Snap
-	opts := &client.SnapOptions{Channel: x.Channel}
-	changeID, err := cli.Refresh(name, opts)
+	updates, err := cli.List(&client.ListOptions{RefreshOnly: true})
```
why are we implementing this in the client? wouldn't it be better to have it in snapd?

chipaca replied:
one possible answer to this question is "because it doesn't seem to fit anywhere". We should talk about that.

mvo5 replied:
TBH I started with this in the client because it was more convenient. It seemed like it made implmenting the wait() simpler. However with a fresh (and less jet-lagged) brain I realize that this is probably not actually true. I think a refreshAll() in the server that just returns a big taskset with all the updates would work just fine for the client as well. I can create a followup branch based on this.

chipaca commented on daemon/api.go:
```diff
@@ -296,6 +296,7 @@ func UserFromRequest(st *state.State, req *http.Request) (*auth.UserState, error
 type metarepo interface {
 	Snap(string, string, store.Authenticator) (*snap.Info, error)
 	FindSnaps(string, string, store.Authenticator) ([]*snap.Info, error)
+	Updates([]string, store.Authenticator) ([]*snap.Info, error)
```
when we switch to the new endpoint we should revisit whether it's updates or refreshes there as well. I think storeside it is still updates; the distinction snapside is that a refresh will (when everything is done) not hit the network, whereas update will. That is: updating is slow, hits the network, downloads stuff, but doesn't (necessarily) refresh. Refresh doesn't (necessarily) hit the network. Klaar?

niemeyer replied:
Refreshing is updating. Let's please not draw a distinction between the two as that will be needless terminology for everybody to learn and keep in mind. If we want to draw a distinction, let's say "refresh --no-network" or something semantically equivalent.

niemeyer replied:
Reminds me of upgrade vs. dist-upgrade.. ugh.

chipaca commented on daemon/api.go:
```diff
@@ -432,6 +480,14 @@ func getSnapsInfo(c *Command, r *http.Request, user *auth.UserState) Response {
 		return InternalError("cannot find route for snaps")
 	}
 
+	if sel, ok := r.URL.Query()["select"]; ok {
```
why not `r.URL.Query().Get("select")`?

mvo5 replied:
*cough* Thanks! Fixed.

chipaca commented:
Mostly :+1:. Some quibbles, but I'd be fine with landing this as a first pass and iterating.

niemeyer commented on client/packages.go:
```diff
@@ -61,10 +62,21 @@ type ResultInfo struct {
 }
 
+type ListOptions struct {
+	RefreshOnly bool
```
Sorry, just a preliminary point before I forget. Will do a more complete review tomorrow.

Can we please call this `SelectRefresh`?  It's almost a nitpick, but it actually enables us to postpone the decision about what it means to mix that with other options (additive or exclusive).

niemeyer commented on client/packages.go:
```diff
@@ -61,10 +62,21 @@ type ResultInfo struct {
 // with names in the given list; if the list is empty, all snaps.
-func (client *Client) ListSnaps(names []string) ([]*Snap, error) {
-	snaps, _, err := client.snapsFromPath("/v2/snaps", nil)
+func (client *Client) List(opts *ListOptions) ([]*Snap, error) {
```
Just to make the obvious case more convenient:
```
if opts == nil {
        opts = &ListOptions{}
}

mvo5 replied:
Done!

niemeyer commented on client/packages_test.go:
```diff
@@ -29,17 +30,29 @@ import (
 
 func (cs *clientSuite) TestClientSnapsCallsEndpoint(c *check.C) {
-	_, _ = cs.cli.ListSnaps(nil)
+	_, _ = cs.cli.List(&client.ListOptions{})
```
... and then we can keep nil here.

niemeyer commented on cmd/snap/cmd_list.go:
```diff
@@ -52,16 +53,31 @@ func (s snapsByName) Less(i, j int) bool { return s[i].Name < s[j].Name }
 	return listSnaps(x.Positional.Snaps)
 }
 
+func listRefreshes() error {
```
Instead of duplicating listSnaps, I suggest passing it a `*client.ListOptions` as an argument.

niemeyer commented on daemon/api.go:
```diff
@@ -419,6 +404,69 @@ func shouldSearchStore(r *http.Request) bool {
 	return false
 }
 
+func storeUpdates(c *Command, r *http.Request, user *auth.UserState) Response {
```
Personal note: continue reviewing from here.

mvo5 commented:
@niemeyer Thanks for the excellent review. I addressed the feedback and I'm especially happy how nice listSnaps(opts *client.ListOptions) now looks :)

niemeyer commented on client/packages.go:
```diff
@@ -61,10 +62,25 @@ type ResultInfo struct {
 }
 
+type ListOptions struct {
+	SelectRefresh bool
```
Sorry for going back and forth here.. can we please just say `Refresh: true` here? If it's good for the command line, it should be good for the Go API.. if it's bad, we should fix both as the CLI is much more important as a user-visible interface.

niemeyer commented on daemon/api.go:
```diff
@@ -419,6 +404,69 @@ func shouldSearchStore(r *http.Request) bool {
+
+	found, err := allLocalSnapInfos(c.d.overlord.State())
+	if err != nil {
+		return InternalError("cannot list local snaps! %v", err)
```
s/!/:/

niemeyer commented on daemon/api.go:
```diff
@@ -419,6 +404,69 @@ func shouldSearchStore(r *http.Request) bool {
+	}
+
+	store := newRemoteRepo()
+	auther, err := c.d.auther(r)
```
We now have the user at hand. This can be:

`auther := user.Authenticator()`

mvo5 replied:
Nice, thank you!

niemeyer commented on daemon/api.go:
```diff
@@ -419,6 +404,69 @@ func shouldSearchStore(r *http.Request) bool {
+
+	// the store gives us everything, we need to client side filter
+	// for the updates
+	allUpdates, err := store.Updates(localSnapsWithChannel, auther)
```
... and can even be inlined here.

niemeyer commented on daemon/api.go:
```diff
@@ -419,6 +404,69 @@ func shouldSearchStore(r *http.Request) bool {
+	updates := []*snap.Info{}
+	for _, update := range allUpdates {
+		local := localSnapMap[update.Name()]
+		if local.Revision > 0 && local.Revision != update.Revision {
```
`!local.Revision.Unset()`

niemeyer commented on debian/snapd.auto-update.timer:
```diff
@@ -0,0 +1,11 @@
+Description=Ubuntu Core Auto Update
+
+[Timer]
+OnCalendar=23,05,11,17:00
```
This should be every day at 4AM, I think..

niemeyer commented on debian/snapd.auto-update.timer:
```diff
@@ -0,0 +1,11 @@
+
+[Timer]
+OnCalendar=23,05,11,17:00
+RandomizedDelaySec=6h
```
Or perhaps at 3AM, with a random delay of 2h?

niemeyer commented:
Looking good. Just a few more comments.

pedronis commented:
I don't think we should make a release based on the old endpoint at this point in time, we need to switch using the new endpoint

pedronis commented on cmd/snap/cmd_snap_op.go:
```diff
@@ -176,29 +176,61 @@ func (x *cmdInstall) Execute([]string) error {
-	name := x.Positional.Snap
-	opts := &client.SnapOptions{Channel: x.Channel}
-	changeID, err := cli.Refresh(name, opts)
+	updates, err := cli.List(&client.ListOptions{Refresh: true})
```
I think we shouldn't ask for the list of updates here in the client actually, we should have a slightly different path in the daemon that instead of doing a search, starts from asking for updates using /metadata (same actually for single updates which means slightly different code in snapstate for update vs installs)

mvo5 replied:
@pedronis Yeah, @chipaca pointed this out too, I was planning to do a followup branch for that, however if you think it is a blocker it can also be done now.

pedronis replied:
@mvo5 not really a blocker, it may be though that with the new endpoint (because it deals with snapids not names, it may be easier the other way), using the new endpoint is a blocker. Not sure. It's something that we need to do soon (I'm happy to help with that once I have progressed a bit on the signing tool stuff)

mvo5 replied:
@pedronis sure, I'm working on the new endpoint as we speak :)

pedronis commented on store/store.go:
```diff
@@ -531,11 +531,37 @@ func (s *SnapUbuntuStoreRepository) FindSnaps(searchTerm string, channel string,
+	for i, inst := range installed {
+		is[i] = metadataInput{
+			SnapID:   inst.SnapID,
+			Channel:  inst.Channel,
```
we have a Channel both on SnapState and the single SideInfos, not sure with rollbacks they are always be in sync, wondering which to pick and wondering whether we shouldn't pass snap.Info but only the precise required data

pedronis commented on store/store.go:
```diff
@@ -531,11 +531,37 @@ func (s *SnapUbuntuStoreRepository) FindSnaps(searchTerm string, channel string,
+		is[i] = metadataInput{
+			SnapID:   inst.SnapID,
+			Channel:  inst.Channel,
+			Revision: inst.Revision.N,
```
this can be negative for sideloaded stuff, I suppose that's ok to pass on

pedronis commented on store/store.go:
```diff
@@ -531,11 +531,37 @@ func (s *SnapUbuntuStoreRepository) FindSnaps(searchTerm string, channel string,
+			Revision: inst.Revision.N,
+			// FIXME: not in snapinfo yet
+			//Confinment: inst.Confinement,
+		}
```
we want to pass epoch too if I understood correctly (probably optionally for now with omitempty)

pedronis commented on daemon/api.go:
```diff
@@ -420,6 +405,65 @@ func shouldSearchStore(r *http.Request) bool {
+		}
+	}
+
+	meta := &Meta{Sources: []string{"updates"}}
```
select=refresh but here we say updates? was this discussed, seems a bit strange

pedronis replied:
@chipaca you worked on the move to select, any input on this?

mvo5 replied:
This was not discussed, I think `refresh` makes more sense here instead of updates. But I'm open for other ideas. Maybe the source is actually just `store`.

pedronis commented on store/store.go:
```diff
@@ -531,11 +531,42 @@ func (s *SnapUbuntuStoreRepository) FindSnaps(searchTerm string, channel string,
+		}
+		is = append(is, metadataInput{
+			SnapID:   inst.SnapID,
+			Channel:  inst.Channel,
```
still wondering where to take channel from? SnapState or SnapState.Current()?  I think SnapState is what all other operations consider atm

mvo5 replied:
@pedronis the current code in the daemons gets all the info via `allLocalSnapInfos()`. Which will use snapstate.Current() to populate this info.

mvo5 replied:
Thinking about this I think we need the snapstate.Channel here.

pedronis replied:
allLocalSnapInfos actually returns the snapstate too (in the aboutSnap structs) fwiw

pedronis commented on store/store_test.go:
```diff
@@ -732,33 +732,34 @@ func (t *remoteRepoTestSuite) TestUbuntuStoreFindAuthFailed(c *C) {
-        "download_sha512": "5364253e4a988f4f5c04380086d542f410455b97d48cc6c69ca2a5877d8aef2a6b2b2f83ec4f688cae61ebc8a6bf2cdbd4dbd8f743f0522fc76540429b79df42"
+const MockUpdatesJSON = `{"_embedded":
+    {"clickindex:package": [
+        {"package_name": "hello-world.canonical",
```
are we really getting back this? this seems very wrong (notice other examples have just hello-world), maybe the curl you are using needs to send release?

pedronis commented on store/store.go:
```diff
@@ -531,11 +531,40 @@ func (s *SnapUbuntuStoreRepository) FindSnaps(searchTerm string, channel string,
+	// build input for the updates endpoint
+	is := make([]metadataInput, 0, len(installed))
+	for _, inst := range installed {
+		if inst.Revision.Local() {
```
is this skipping here what we want? it seems the wrong place to decide this,  for example we decided that "snap refresh name" would get us out of trymode for example

mvo5 replied:
Hm, good point. I think you are right, this is not the right place. I was thinking about leaking data to the store but I guess I was over-thinking it and to get out of a local install back to the store it is needed.

pedronis commented on store/store.go:
```diff
@@ -531,11 +531,37 @@ func (s *SnapUbuntuStoreRepository) FindSnaps(searchTerm string, channel string,
+}
+
+// Updates returns the available updates for a list of snap identified by fullname with channel.
+func (s *SnapUbuntuStoreRepository) Updates(installed []*snap.Info, auther Authenticator) (snaps []*snap.Info, err error) {
```
as we discussed in the standup this probably should take some variations of metadataInput (exported), with Info doesns't match exactly what we need and makes it tempting to take decisions here

pedronis commented on store/store.go:
```diff
@@ -562,12 +562,16 @@ func (s *SnapUbuntuStoreRepository) Updates(installed []*CurrentSnap, auther Aut
 	currentSnaps := make([]currentSnapJson, len(installed))
 	for i, cs := range installed {
+		revision := cs.Revision.N
+		if cs.Revision.Local() {
```
I think I would write it as !cs.Revision.Store() ?

pedronis replied:
btw the reasoning for this seems to be, hide it from the store if it's not a store revision, not sure we will invent more strange revision variants, but seems safer

mvo5 replied:
Indeed, good suggestion!

pedronis commented:
this looks good to me now, as I said later we want to make calling /metadata a concern of snapstate but is a good start, I understand we still need to work out if the results are what we expect from the server

and this probably needs another review pass by somebody else given that it has changed quite a bit again

mvo5 commented:
retest this please

mvo5 commented:
@pedronis Thanks a bunch for your excellent feedback to move this branch forward. I agree that we want to move the logic to the /metadata calling into snapstate. And :+1: for another review.

niemeyer commented:
Re-reviewing...

niemeyer commented on daemon/api.go:
```diff
@@ -420,6 +405,72 @@ func shouldSearchStore(r *http.Request) bool {
+	localSnapsInfo := make([]*store.CurrentSnap, len(found))
+	localSnapMap := map[string]*snap.Info{}
+	for i, sn := range found {
+		localSnapsInfo[i] = &store.CurrentSnap{
```
s/CurrentSnap/RefreshCandidate/?

Sounds like a misnomer to name it current with details which do not match the current snap (channel). We'll likely have that situation happening in other cases soon as well (--devmode on the CLI may override the status quo).

niemeyer commented on daemon/api.go:
```diff
@@ -420,6 +405,72 @@ func shouldSearchStore(r *http.Request) bool {
+			SnapID:      sn.info.SnapID,
+			Revision:    sn.info.Revision,
+			Epoch:       sn.info.Epoch,
+			Confinement: sn.info.Confinement,
```
This is similar to channel, I think. We care more about what the current situation is (snapst) than what the snap claims its preference to be. The snap may be strictly confined, but we may have it in devmode locally, in which case updating it to another devmode snap would be fine.

pedronis replied:
we should also keep in mind that we don't want things in try to be refreshed except if they are manually requested to be. not super clear where that decision should happen though, it would still be good if snap refresh --list told the user what will happen if invoking snap refresh try-snap

niemeyer replied:
True, and I think it should happen right here.  If sn.snapst is in TryMode, we just don't put that in the candidate list.

niemeyer commented on daemon/api.go:
```diff
@@ -420,6 +405,72 @@ func shouldSearchStore(r *http.Request) bool {
+			Epoch:       sn.info.Epoch,
+			Confinement: sn.info.Confinement,
+		}
+		localSnapMap[sn.info.Name()] = sn.info
```
s/localSnapMap/candidateMap/, potentially.

niemeyer commented on daemon/api.go:
```diff
@@ -420,6 +405,72 @@ func shouldSearchStore(r *http.Request) bool {
+	// the store gives us everything, we need to client side filter
+	// for the updates
+	store := newRemoteRepo()
+	allUpdates, err := store.Updates(localSnapsInfo, user.Authenticator())
```
It'd be nice if we could attempt to align terminology after the "refresh" concept, at least inside snapd itself (I reckon we cannot easily change store-side naming). So, this might be ListRefresh or similar, for example.

niemeyer commented on daemon/api.go:
```diff
@@ -420,6 +405,72 @@ func shouldSearchStore(r *http.Request) bool {
+	updates := []*snap.Info{}
+	for _, update := range allUpdates {
+		local := localSnapMap[update.Name()]
+		if !local.Revision.Unset() && local.Revision != update.Revision {
```
Why is that not being done by the store method itself?  We explicitly asked for the refreshes, and it has both revisions internally available. It seems unexpected and error prone to have it returning the exact same revision we handed in.

mvo5 replied:
Indeed, thanks for this! Updated.

niemeyer commented on daemon/api.go:
```diff
@@ -420,6 +405,72 @@ func shouldSearchStore(r *http.Request) bool {
+		}
+	}
+
+	meta := &Meta{Sources: []string{"refresh"}}
```
I think `Sources` should die after @chipaca's work on splitting `/v2/snaps` from `/v2/find` lands, as it means all responses have the expected sources. Can we please drop this?

niemeyer commented on daemon/api.go:
```diff
@@ -433,6 +484,10 @@ func getSnapsInfo(c *Command, r *http.Request, user *auth.UserState) Response {
 	}
 
+	if r.URL.Query().Get("select") == "refresh" {
+		return storeUpdates(c, r, user)
```
Loose thought for the future: we may eventually have something similar to what we had before, with CandidateRevision pre-cached, and possibly pre-downloaded.

niemeyer commented on debian/snapd.auto-update.service:
```diff
@@ -0,0 +1,8 @@
+[Unit]
+Description=Ubuntu Core Auto Update
```
`Automatically refresh installed snaps`

?

niemeyer replied:
The file names should change too. I'd probably drop the "auto-" prefix altogether here, keeping just `snapd.refresh.service` and `snapd.refresh.timer`.. the fact it's automatic is implicit from context.

niemeyer commented on debian/snapd.auto-update.timer:
```diff
@@ -0,0 +1,11 @@
+[Unit]
+Description=Ubuntu Core Auto Update
```
`Timer to automatically refresh installed snaps`

niemeyer commented on docs/rest.md:
```diff
@@ -247,6 +247,14 @@ Which channel to search in.
+#### `select`
+
+Filter from the given selection. Currently only limiting to refreshable
+snaps is supported via the `refresh` key.
```
Idle thought (please don't kill me): throughout the review I've been wondering if this is best suited for list (/v2/snaps) or for find (/v2/find). The semantics would be exactly the same and all: same output (meaning, `find --refresh` would return what you have in this branch already), similar Client interface (`cli.Find(&FindOptions{Refresh: true})`), same API (`select=refresh`), etc. The only reason for the change is that `/v2/find` has been about looking for things in the store that we don't have at hand, while `/v2/snaps` became the way to talk about what we have locally. If we draw the line that way, it's awkward to have an option that changes that behavior and suddenly makes /v2/snaps pause.

niemeyer replied:
Another motivator for this: `select=refresh` is a bit of a misnomer. It's not filtering.. it's changing the result set completely.

As an alternative for the CLI if `snap find --refresh` sounds awkward, we might go full circle into your proposal of `snap refresh --list`. Internally, that would might still map to `cli.Find` per idea above. This might be sensible as the output would look more like `list` than `find`, and killing the `--list` option exercises what was just exposed.

Apologies again. I realize this has seen a long process already, and I feel bad for discussing it further, but it's probably wise to spend a bit more time on this if we feel it's worth it.

What do you think? @chipaca? @pedronis?

pedronis replied:
I think I agree that this belongs more to find at the REST API level if we want to maintain that /snaps doesn't need to talk remotely, otoh I don't like snap find --refresh that much as command line way to get to it

so snap refresh --list would seem again better

mvo5 replied:
I changed to `snap refresh --list` and updated the rest API so that we have a `select: refresh` on the find endpoint now.

niemeyer commented on integration-tests/tests/list_test.go:
```diff
@@ -64,3 +68,38 @@ func (s *listSuite) TestListMustPrintAppVersion(c *check.C) {
 	c.Assert(listOutput, check.Matches, expected)
 }
+
+func (s *listSuite) TestListRefreshesMustPrintUpdates(c *check.C) {
```
`TestListRefresh`? :)

niemeyer commented on integration-tests/tests/snap_update_all_test.go:
```diff
@@ -0,0 +1,69 @@
+	common.SnappySuite
+}
+
+func (s *snapRefreshAllSuite) TestAllUpdate(c *check.C) {
```
s/update/refresh/ on file name and file content.

niemeyer commented on integration-tests/tests/snap_update_app_test.go:
```diff
@@ -67,6 +67,6 @@ func (s *snapRefreshAppSuite) TestAppUpdate(c *check.C) {
 
 	// run the fake update
-	output := updates.CallFakeSnapRefresh(c, snap, updates.NoOp, fakeStore)
+	output := updates.CallFakeSnapRefreshForSnap(c, snap, updates.NoOp, fakeStore)
```
s/update/refresh/ on filename.

niemeyer commented on integration-tests/testutils/updates/updates.go:
```diff
@@ -48,48 +45,31 @@ func NoOp(snapPath string) error {
 }
 
-func makeFakeUpdateForSnap(c *check.C, snap, targetDir string, changeFunc ChangeFakeUpdateSnap) error {
+func MakeFakeUpdateForSnap(c *check.C, snap, targetDir string, changeFunc ChangeFakeUpdateSnap) error {
```
s/update/refresh/ on filename and content.

niemeyer commented on store/store.go:
```diff
@@ -531,11 +531,55 @@ func (s *SnapUbuntuStoreRepository) FindSnaps(searchTerm string, channel string,
+	Epoch       string
+	Confinement snap.ConfinementType
+
+	// the desired channel
```
Renaming to RefreshCandidate should avoid the need to state this everywhere.

niemeyer commented on store/store.go:
```diff
@@ -531,11 +531,55 @@ func (s *SnapUbuntuStoreRepository) FindSnaps(searchTerm string, channel string,
+	SnapID      string
+	Revision    snap.Revision
+	Epoch       string
+	Confinement snap.ConfinementType
```
Should probably be DevMode, per earlier comment, and reflect the current state rather than the snap opinion.

mvo5 commented:
Thanks for the excellent and thorough review. I addressed the issues raised. Please let me know if there is anything else that needs t be done. I also checked the new store API against staging and it seems to be working as expected.

matiasb commented on store/store.go:
```diff
@@ -531,11 +531,57 @@ func (s *SnapUbuntuStoreRepository) FindSnaps(searchTerm string, channel string,
+	Channel  string `json:"channel"`
+	Revision int    `json:"revision,omitempty"`
+	Epoch    string `json:"epoch"`
+	DevMode  bool   `json:"devmode"`
```
I do understand you need to let the store know the confinement switch was used when the snap was installed, and that it could be different from the confinement reported by the revision of that snap in the store, but the client could still send the `confinement` parameter with the actual (installation) value, correct?

We agree the store needs to know how the revision was actually installed to decide if there is/which update to return. My concern is that if eventually there is a new possible confinement value, having a boolean flag doesn't seem enough, meanwhile having the confinement field will always report the confinement the revision has been installed with, and the store will have the information to decide what to do accordingly. Also, at least from the store perspective, 'devmode' is a particular value for confinement so it looks odd having it as a key.

niemeyer replied:
@matiasb So far we've been attempting to have the confinement setting reflecting one single concept: what is the preferred mode of confinement operation by the snap, which is related to but not the same as the current state of the snap. You are right that this confinement setting can have three or more values, but at the values we considered possible so far, such as "enforced", would mean we want the `devmode` to necessarily be false for the current state, rather than being completely orthogonal to it.

There's a small chance we'll have indeed a third mode which is completely orthogonal, but this will break more concepts than simply this API. The `devmode` term is used throughout snapd, including CLI, API, and internal state.

matiasb replied:
@niemeyer Yeah, I understand that snapd internally uses confinement and devmode this way, but somehow I feel this change makes the store need to be aware of that too. The store doesn't know anything about a devmode, but it sure should be able to understand confinements and decide which the right update (if any) for a given a snap setup description (ie. channel, confinement, epoch) is. The client could decide whatever value to pass for each.
AFAICT, the store could just return the expected response for the requested scenario (without knowing anything else but the data that it already stores and indexes from the snaps metadata).
That's what keeps me doubting from changing the original decided API. Makes sense?

niemeyer replied:
It doesn't make much sense to me particularly, because you're in fact asking to send something that is _not_ the confinement property for the snap, but whether the snap was installed in devmode or not.

That said, I don't mind much to be honest. We can very easily map the store setting to whatever the store wants us to name the setting as.

@mvo5 if the store expects this to be named as `confinement`, let's please change only in `currentSnapJson` and set it appropriately with strict or devmode depending on what we find in `candidate.DevMode`.

niemeyer commented:
retest this please

mvo5 commented:
The store deployed the new endpoint now:
```
$ curl -s --data-binary '{"snaps":[{"snap_id":"buPKUD3TKqCOgLEjjHx5kSiCpIs5cMuQ","channel":"stable","revision":0,"confinement":"strict"}],"fields":["snap_id","package_name","revision","version","download_url"]}'  -H 'content-type: application/json' https://search.apps.ubuntu.com/api/v1/metadata|python -m json.tool
{
    "_embedded": {
        "clickindex:package": [
            {
                "_links": {
                    "self": {
                        "href": "https://search.apps.ubuntu.com/api/v1/package/buPKUD3TKqCOgLEjjHx5kSiCpIs5cMuQ"
                    }
                },
                "download_url": "https://public.apps.ubuntu.com/download-snap/buPKUD3TKqCOgLEjjHx5kSiCpIs5cMuQ_25.snap",
                "package_name": "hello-world",
                "revision": 25,
                "snap_id": "buPKUD3TKqCOgLEjjHx5kSiCpIs5cMuQ",
                "version": "6.0"
            }
        ]
    },
    "_links": {
        "curies": [
            {
                "href": "https://wiki.ubuntu.com/AppStore/Interfaces/ClickPackageIndex#reltype_{rel}",
                "name": "clickindex",
                "templated": true
            }
        ]
    }
}
```

pedronis commented:
@mvo5 @matiasb are we sending devmode? or confinement to the store? (I think the last reluctanct agreement was to send confiment (but based on devmode flag) ?

pedronis commented on store/store.go:
```diff
@@ -531,11 +531,67 @@ func (s *SnapUbuntuStoreRepository) FindSnaps(searchTerm string, channel string,
+
+	// The store expects a "confinement" value {"strict", "devmode"}.
+	// We map this accordingly from our devmode bool, we do not
+	// use the value of the current snap as we are interessted in the
```
typo: interested

pedronis commented on store/store.go:
```diff
@@ -555,15 +611,20 @@ func (s *SnapUbuntuStoreRepository) Updates(installed []string, auther Authentic
+	res := make([]*snap.Info, 0, len(updateData.Payload.Packages))
+	for _, rsnap := range updateData.Payload.Packages {
+		// the store also gives us identical revisions, filter those
+		// out, we are not interessted
```
same

pedronis commented:
lgtm, just couple of typos in the comments

chipaca commented:
retest this please

zyga commented on client/packages.go:
```diff
@@ -67,9 +68,14 @@ type ResultInfo struct {
 	SuggestedCurrency string `json:"suggested-currency"`
 }
 
```
A comment would be nice here

zyga commented on client/packages.go:
```diff
@@ -67,9 +68,14 @@ type ResultInfo struct {
+	Query   string
+}
+
 // ListSnaps returns the list of all snaps installed on the system
```
s/ListSnaps/List/

zyga commented on cmd/snap/cmd_snap_op.go:
```diff
@@ -188,27 +188,60 @@ func (x *cmdInstall) Execute([]string) error {
 	Channel    string `long:"channel" description:"Refresh to the latest on this channel, and track this channel henceforth"`
 	Positional struct {
 		Snap string `positional-arg-name:"<snap>"`
-	} `positional-args:"yes" required:"yes"`
```
\o/ :-) (just happy that this is done :-)

zyga commented on debian/rules:
```diff
@@ -27,15 +27,15 @@ override_dh_systemd_enable:
 		--no-enable \
 		-psnapd \
-		snappy-autopilot.service
+		snappy-auto-update.service
```
Do we want to call this `snapd-auto-refresh` instead?

niemeyer replied:
Yes, this should be snapd.refresh.service and snapd.refresh.timer, I think.

zyga commented:
One important question on debhelper / systemd discrepancy. Apart from that a few nitpicks. LGTM assuming that either I'm wrong or the packaging is updated to mathc.

niemeyer commented on client/packages.go:
```diff
@@ -96,10 +102,16 @@ func (client *Client) ListSnaps(names []string) ([]*Snap, error) {
 // store for this system and that match the query
-func (client *Client) FindSnaps(query string) ([]*Snap, *ResultInfo, error) {
-	q := url.Values{}
+func (client *Client) FindSnaps(opts *FindOptions) ([]*Snap, *ResultInfo, error) {
```
s/FindSnaps/Find/

We should change List to become similar too (ListOptions, and ResultInfo), but we can do that independently.

niemeyer commented on daemon/api.go:
```diff
@@ -362,8 +363,11 @@ func searchStore(c *Command, r *http.Request, user *auth.UserState) Response {
+		return storeUpdates(c, r)
+	}
 
+	query := r.URL.Query()
```
This may be moved above the conditional block so we can already use it there.

Also, is there an easy way to respect "q", given that we're asking the store for these updates?

If not, we need to reject queries that provide both, rather than silently doing something else.

niemeyer commented on daemon/api.go:
```diff
@@ -421,6 +409,71 @@ func shouldSearchStore(r *http.Request) bool {
+			continue
+		}
+
+		// get confinment preference from the snapstate
```
I'm glad I'm not the only one that does that exact typo. Confinement is typing-unfriendly. :)

niemeyer commented on daemon/api.go:
```diff
@@ -421,6 +409,71 @@ func shouldSearchStore(r *http.Request) bool {
+		})
+	}
+
+	auther, err := c.d.auther(r)
```
searchStore should send the user in so we can do:
```
auther := user.Authenticator()
```
We should probably kill the auther helper to avoid redoing that historical use.

niemeyer commented on daemon/api.go:
```diff
@@ -421,6 +409,71 @@ func shouldSearchStore(r *http.Request) bool {
+		return InternalError("%v", err)
+	}
+
+	store := newRemoteRepo()
```
Unrelated to this branch: it's strange we instantiate that all the time. We should really keep it around in the daemon.

niemeyer commented on daemon/api.go:
```diff
@@ -421,6 +409,71 @@ func shouldSearchStore(r *http.Request) bool {
+}
+
+func sendStorePackages(route *mux.Route, meta *Meta, found []*snap.Info) Response {
+	results := make([]*json.RawMessage, len(found))
```
This needs to be `make([]*json.RawMessage, 0, len(found))`, and we need to use `append` below, due to the `continue`.

niemeyer commented on store/store.go:
```diff
@@ -531,11 +531,67 @@ func (s *SnapUbuntuStoreRepository) FindSnaps(searchTerm string, channel string,
+	// The store expects a "confinement" value {"strict", "devmode"}.
+	// We map this accordingly from our devmode bool, we do not
+	// use the value of the current snap as we are interested in the
+	// users intention, not the actual value of the snap itself.
```
Nice comment, thank you.

niemeyer commented:
Last review, I suspect. LGTM once you're happy about these.
